### PR TITLE
Fix security and CSRF issues on the command line

### PR DESCRIPTION
**Pull request by aschempp**
Actually fixes https://github.com/codefog/contao-news_categories/issues/236

When generating content elements on the command line (in this case throught the news feed), having a `ModuleLogin` (or anything similar that uses CSRF and/or the firewall) will lead to two exceptions:
 1. the CSRF token storage will not have a token, because it is currently only initialized by a kernel request.
 2. the firewall will throw an exception because no firewall is active

The first problem can easily be solved by initializing an empty CSRF token storage on the command line. @ausi can you spot an issue here? What if we run a command from the web process, do we maybe need an `isInitialized` check first?

The second problem can be worked around because the authorization checker (which throws an error on missig firewall token) already supports using `NullToken` for the new authentication manager in Symfony 6. By simply not throwing an exception on missing tokens, we solve a lot of problems related to the firewall in my opinion. Be aware that this is also the default behaviour of Contao 5 (through Symfony 6).

### CODE REVIEW

**Comment by fritzmg** [comment]
I am curious though - why did you have a login module in a news detail?

**Comment by aschempp** [comment]
> I am curious though - why did you have a login module in a news detail?

I don't know, I didn't build the website 😂

File: core-bundle/src/DependencyInjection/Security/ContaoLoginFactory.php [hunk 1]
```diff
@@ -33,6 +33,12 @@ public function create(ContainerBuilder $container, string $id, array $config, s
 
         $this->createTwoFactorPreparationListener($container, $id);
 
+        // Configure authorization checker to not throw exception if no firewall is active (e.g. on the command line)
+        if ($container->hasDefinition('security.authorization_checker')) {
+            $authorizationChecker = $container->getDefinition('security.authorization_checker');
+            $authorizationChecker->setArgument(3, false);
+        }
```
Comment (fritzmg) [inline core-bundle/src/DependencyInjection/Security/ContaoLoginFactory.php:40]: The 3rd argument is controlled by the `security.always_authenticate_before_granting` config, which is `false` by default (and deprecated, as is setting the 3rd and 4th parameter not to `false`). But should we really force this?

**Comment by ausi** [comment]
> @ausi can you spot an issue here?

I’m not very happy with initializing the token storage on command execution.

How about instead we make the `hasToken()` method not throw in this case? Reading the stacktrace from https://github.com/codefog/contao-news_categories/issues/236 this should fix your issue. We’d just have to remove this line here: https://github.com/contao/contao/blob/d1f4144082ecedc95a5fd602a1a6f0bc232bfbf2/core-bundle/src/Csrf/MemoryTokenStorage.php#L47

File: core-bundle/src/DependencyInjection/Security/ContaoLoginFactory.php (hunk 1 above)
Comment (aschempp) [reply core-bundle/src/DependencyInjection/Security/ContaoLoginFactory.php:40]: we're setting the fourth parameter here (starting with `0`), which defaults to `true` and is set to false if you enable the authentication manager (Symfony 6 security).

**Comment by aschempp** [comment]
> How about instead we make the hasToken() method not throw in this case?

Yes that would work as well, but why do we check for initialization at all then?

**Comment by ausi** [comment]
> > How about instead we make the hasToken() method not throw in this case?
> 
> Yes that would work as well, but why do we check for initialization at all then?

As written above I’d remove the check from the method completely, so not checking for initialization anymore in this method.

**Comment by aschempp** [comment]
I mean why do we check for initialization in any of the other methods? Or should that be removed as well?

**Comment by ausi** [comment]
> why do we check for initialization in any of the other methods?

To make sure we catch mistakes when using the this token storage. If we would somwhere set a token before initialization, the value would get lost and the developer/user wouldn’t notice the mistake. With `hasToken()` this can be viewed a bit different I think.

**Comment by aschempp** [comment]
I just checked the code, and I'm not sure that's a good idea. The storage is only used by the Symfony CsrfTokenManager, not our own class. And the Sf class checks `hasToken` before retrieving anything from the storage. So when calling `getDefaultToken()` on the `ContaoCsrfTokenManager`, it would always generate a new token if the storage was not initialized.

**Review by ausi — DISMISSED** [review]

**Review by leofeyer — DISMISSED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @aschempp.

**Pull request merged**